File: Javascript/PalindromePartitioning.js
// Leetcode Problem - 131

var partition = function(s){
    const result = []
    const part = []

    function dfs(i){
        if (i>=s.length){
            result.push([...part])
            return
        }

        for (let j=i;j<s.length;j++){
            if (isPalindrome(s,i,j)){
                part.push(s.slice(i,j+1))
                dfs(j+1)
                part.pop()
            }
        }
    }

    dfs(0)

    function isPalindrome(s,l,r){
        while (l<r){
            if (s[l]!==s[r]){
                return false
            }
            l+=1
            r-=1
        }
        return true
    }

    return result
}

s = "aab"
console.log(partition(s))

File: Python/PalindromePartitioning.py
## Leetcode Problem - 131

def partition(s):
    result = []
    part = []

    def isPalindrome(s,l,r):
        while l<r:
            if s[l]!=s[r]:
                return False
            l+=1
            r-=1
            
        return True

    def dfs(i):
        if i >=len(s):
            result.append(part.copy())
            return
        
        for j in range(i,len(s)):
            if isPalindrome(s,i,j):
                part.append(s[i:j+1])
                dfs(j+1)
                part.pop()

    dfs(0)


s = "aab"
partition(s)